File: 63/unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, grid: List[List[int]]) -> int:
        M, N = len(grid), len(grid[0])
        table = [[0]*N for _ in range(M)]

        for i in range(M):
            for j in range(N):
                if grid[i][j]==1: table[i][j] = 0
                elif i==0 and j==0: table[i][j] = 1
                elif i==0: table[i][j] = table[i][j-1]
                elif j==0: table[i][j] = table[i-1][j]
                else: table[i][j] = table[i-1][j] + table[i][j-1]
        return table[M-1][N-1]

File: 63/unique-paths-ii.cs
public class Solution {
    public int UniquePathsWithObstacles(int[][] grid) {
        if(grid[0][0]==1) return 0; //最左上角出發點卡住，就沒辦法走

        //C# 用 .Length 而 Java 用 .length 而 C++ 用 ArrayList size()
        int M = grid.Length, N = grid[0].Length;
        int [,]table = new int[M,N];
        //C# 的規則陣列，是用 int[,] table = new int[3,4];
        //不規則陣列，是 int[][] table = new int[3][]; 再逐一new
        // table[0] = new int[4]; 
        // table[1] = new int[4]; 
        // table[2] = new int[4]
        //https://ithelp.ithome.com.tw/articles/10213228


        table[0,0] = 1; //出發點有1個可能
        for(int i=0; i<M; i++){
            for(int j=0; j<N; j++){
                if(grid[i][j]==1) table[i,j] = 0;
                else{
                    if(i!=0) table[i,j] = table[i-1,j];
                    if(j!=0) table[i,j] += table[i,j-1];
                }
            }
        }
        return table[M-1,N-1];
    }
}
